src/Npgsql/NpgsqlCommand.cs

```diff
@@ -813,7 +867,7 @@ await connector.DescribeMessage
             CleanupSend();
         }
 
-        async Task SendPrepare(bool async)
+        async Task SendParseDescribe(bool async, bool prepare)
```

Comment:
I finally do like that you've managed to combine the prepare and derivation paths into the same function, but here a thought...

Rather than having a parameter controlling preparation, why not just check on each statement in the loop below whether has an associated `PreparedStatement`, and use that to decide what to do. After all, the only differences between prepare and derive is the statement name and whether parameter typecodes are sent in the Parse ("preSpecifiedInputParameters"). This should remove some of the conditions and reduce the code size.

Reply:
I don't agree here.
`Prepare()` and `DeriveParameters()` are both public interface methods. It seems perfectly valid to me that someone would first call `Prepare()` and afterwards `DeriveParameters()` for the same `NpgsqlCommand` and expect no interference.
In this case `SendParseDescribe()` would find an associated `PreparedStatement` and try reprepare it (I didn't check it but this would probably cause an exception) instead of just deriving parameters using the unnamed prepared statement.

Reply:
> It seems perfectly valid to me that someone would first call Prepare() and afterwards DeriveParameters() for the same NpgsqlCommand and expect no interference.

That's quite an odd scenario - Prepare() requires explicitly specified parameters, so calling DeriveParameters() after Prepare() seems quite useless, no? Do you have a scenario in mind? It is in fact somewhat dangerous, since the derivation process *may* derive different parameter types from what the user explicitly set before preparing, and then executing the prepared command could fail...

In a way, the same is true even for auto-prepared statements - if you've executed the statement so that it was auto-prepared, you've been providing parameter values (and types) and derivation makes no real sense anymore.

So this is connected to my general question above on derivation and preparation. I don't see a point in derivation after preparation (or even after executing) - although maybe you have a scenario in mind. If there's no point for it, I'd rather throw an exception if derivation is attempted on an explicitly prepared command, and silently detach an auto-prepared command if derivation is called for it. This would ensure that we only ever derive on totally unprepared commands.

What do you think?

Reply:
I kind of agree. Throwing an exception makes sense too - possibly even more.
I suggest I check how SqlCommand/SqlCommandBuilder behave and be consistent with it.
Ok?

Reply:
@roji I checked SqlClient behaviour today.
SqlCommandBuilder doesn't support DeriveParameters() for a SqlCommand with CommandType.Text and throws an exception in this case but it supports the sequence Prepare(); DeriveParameters() for commands with CommandType.StoredProcedure.
Yet the documentation states "If CommandType is set to StoredProcedure, the call to Prepare should succeed, although it may cause a no-op." which certainly makes sense since the preparation of a stored procedure seems pretty nonsensical.
I think compatibility isn't an issue unless Microsoft decides to support DeriveParameters() for a SqlCommand with CommandType.Text and implements it in the same way, it is implemented for CommandType.StoredProcedure.
Anyways, I don't really care that much how we solve this.
If you still think an exception is the way to go, I'll implement it that way.

Reply:
Thanks for checking the SqlClient behavior... IIRC Prepare() is a no-op in all recent versions of SQLServer, which manages caching transparently server-side for all statements (i.e. it's all implicit rather than requiring explicit preparation like PostgreSQL).

So yes, let's throw an exception if an explicitly-prepared command is derived.

Reply:
Looking at the code again, after the above discussion I again don't see a reason for `SendParseDescribe()` to have a `prepare` parameter - the code can simply branch on whether whether `statement` has an attached `PreparedStatement`, and populate the ParseMessage based on that (statement name and input parameters). That should simplify the code.